org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/CompletionProposalReplacementProvider.java

```diff
@@ -111,31 +127,31 @@ public void updateReplacement(CompletionProposal proposal, CompletionItem item,
 		StringBuilder completionBuffer = new StringBuilder();
 		Range range = null;
 		if (isSupportingRequiredProposals(proposal)) {
-			CompletionProposal[] requiredProposals= proposal.getRequiredProposals();
+			CompletionProposal[] requiredProposals = proposal.getRequiredProposals();
 			if (requiredProposals != null) {
 				for (CompletionProposal requiredProposal : requiredProposals) {
-					switch(requiredProposal.getKind()) {
-					case CompletionProposal.TYPE_IMPORT:
-					case CompletionProposal.METHOD_IMPORT:
-					case CompletionProposal.FIELD_IMPORT:
-						appendImportProposal(completionBuffer, requiredProposal, proposal.getKind());
-						break;
-					case CompletionProposal.TYPE_REF:
-						org.eclipse.lsp4j.TextEdit edit = toRequiredTypeEdit(requiredProposal, trigger, proposal.canUseDiamond(context));
+					switch (requiredProposal.getKind()) {
+						case CompletionProposal.TYPE_IMPORT:
+						case CompletionProposal.METHOD_IMPORT:
+						case CompletionProposal.FIELD_IMPORT:
+							appendImportProposal(completionBuffer, requiredProposal, proposal.getKind());
+							break;
+						case CompletionProposal.TYPE_REF:
+							org.eclipse.lsp4j.TextEdit edit = toRequiredTypeEdit(requiredProposal, trigger, proposal.canUseDiamond(context));
 							if (proposal.getKind() == CompletionProposal.CONSTRUCTOR_INVOCATION || proposal.getKind() == CompletionProposal.ANONYMOUS_CLASS_CONSTRUCTOR_INVOCATION
 									|| proposal.getKind() == CompletionProposal.ANONYMOUS_CLASS_DECLARATION) {
-							completionBuffer.append(edit.getNewText());
-							range = edit.getRange();
-						} else {
-							additionalTextEdits.add(edit);
-						}
-						break;
-					default:
-						/*
-						 * In 3.3 we only support the above required proposals, see
-						 * CompletionProposal#getRequiredProposals()
-						 */
-						Assert.isTrue(false);
+								completionBuffer.append(edit.getNewText());
+								range = edit.getRange();
+							} else {
+								additionalTextEdits.add(edit);
+							}
```

Comment:
Not sure if we should do something like:
```java
if (proposal.getKind() == CompletionProposal.CONSTRUCTOR_INVOCATION || proposal.getKind() == CompletionProposal.ANONYMOUS_CLASS_CONSTRUCTOR_INVOCATION
		|| proposal.getKind() == CompletionProposal.ANONYMOUS_CLASS_DECLARATION) {
	org.eclipse.lsp4j.TextEdit edit = toRequiredTypeEdit(requiredProposal, trigger, proposal.canUseDiamond(context));
	completionBuffer.append(edit.getNewText());
	range = edit.getRange();
} else if (addImports) {
	org.eclipse.lsp4j.TextEdit edit = toRequiredTypeEdit(requiredProposal, trigger, proposal.canUseDiamond(context));
	additionalTextEdits.add(edit);
}
```
are all additional edits imports?

Reply:
> are all additional edits imports?

No. In most cases they are imports. But there is a special case [CompletionHandlerTest.java#testCompletion_method_withLSPV3()](https://github.com/eclipse/eclipse.jdt.ls/blob/c2bdb489d305ece226da4e55f6e349d65ce34383/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/CompletionHandlerTest.java#L685-L686), besides the imports, there is another additonal edit to replace "HashMap" with "HashMap". It look confusing (redudant), but after playing for a while, I found a case that might better explain it.

```java
public class A {
    public static void main(String[] args) {
        HashMap map = new HashMap<>();
        map.putA
    }
}

```

```json
// additional edits for `putAll`
    "additionalTextEdits": [
        {
            "range": {
                "start": {
                    "line": 2,
                    "character": 8
                },
                "end": {
                    "line": 2,
                    "character": 15
                }
            },
            "newText": "HashMap<K,V>" //<-- this is the `toRequiredTypeEdit` one
        },
        {
            "range": {
                "start": {
                    "line": 0,
                    "character": 0
                },
                "end": {
                    "line": 0,
                    "character": 0
                }
            },
            "newText": "import java.util.HashMap;\n\n"
        }
```


No matter whether they are imports or not, the additional edits will be correctly calculated in the resolve stage. The change you proposed is surely harmless, and can also reduce the times of calling  `toRequiredTypeEdit`.